packages/xforms-engine/src/lib/reactivity/validation/createValidation.ts

```diff
@@ -0,0 +1,196 @@
+import type { Accessor } from 'solid-js';
+import { createMemo } from 'solid-js';
+import type { OpaqueReactiveObjectFactory } from '../../../client/OpaqueReactiveObjectFactory.ts';
+import type {
+	TextRange as ClientTextRange,
+	ValidationTextRole,
+} from '../../../client/TextRange.ts';
+import { VALIDATION_TEXT } from '../../../client/constants.ts';
+import type {
+	AnyViolation,
+	ConditionSatisfied,
+	ConditionValidation,
+	ConditionViolation,
+	ValidationCondition,
+} from '../../../client/validation.ts';
+import type { ValidationContext } from '../../../instance/internal-api/ValidationContext.ts';
+import { TextChunk } from '../../../instance/text/TextChunk.ts';
+import { TextRange } from '../../../instance/text/TextRange.ts';
+import type { MessageDefinition } from '../../../parse/text/MessageDefinition.ts';
+import { createComputedExpression } from '../createComputedExpression.ts';
+import type {
+	SharedNodeState,
+	SharedNodeStateOptions,
+} from '../node-state/createSharedNodeState.ts';
+import { createSharedNodeState } from '../node-state/createSharedNodeState.ts';
+import type { ReactiveScope } from '../scope.ts';
+import { createTextRange } from '../text/createTextRange.ts';
+
+type EngineViolationMessage<Role extends ValidationTextRole> = ClientTextRange<Role, 'engine'>;
+
+const engineViolationMessage = <Role extends ValidationTextRole>(
+	context: ValidationContext,
+	role: Role
+): Accessor<EngineViolationMessage<Role>> => {
+	const messageText = VALIDATION_TEXT[role];
+	const chunk = new TextChunk(context.root, 'static', messageText);
+	const message = new TextRange('engine', role, [chunk]);
+
+	return () => message;
+};
+
+const createViolationMessage = <Role extends ValidationTextRole>(
+	context: ValidationContext,
+	role: Role,
+	definition: MessageDefinition<Role> | null
+) => {
+	if (definition == null) {
+		return engineViolationMessage(context, role);
+	}
+
+	return createTextRange(context, role, definition);
+};
+
+// prettier-ignore
+type ComputedConditionValidation<
+	Condition extends ValidationCondition
+> = Accessor<ConditionValidation<Condition>>;
+
+const constraintValid = (): ConditionSatisfied<'constraint'> => {
+	return {
+		condition: 'constraint',
+		valid: true,
+		message: null,
+	};
+};
+
+const createConstraintValidation = (
+	context: ValidationContext
+): ComputedConditionValidation<'constraint'> => {
+	return context.scope.runTask(() => {
+		const { constraint, constraintMsg } = context.definition.bind;
+
+		if (constraint == null) {
+			return constraintValid;
+		}
+
+		const isValid = createComputedExpression(context, constraint, {
+			arbitraryDependencies: [context],
+		});
```

Comment:
🤔 if I understand this correctly, we are watching/subscribing to the complete tree here - any change in any part of the Form will trigger this, right?

Reply:
No, but an understandable misunderstanding (as a side effect of specifying a limited interface for `context`)! 🙃 

What this does is subscribe to the node (`StringField`, `SelectField`) itself. This ensures that validation is updated when a node's value changes. This will often be the case already for `constraint` (as it's common to reference the value in the `constraint` expression), but it's not guaranteed. It generally won't be the case for `required`. Without this subscription, validation state would be inconsistent with the node's `value` state unless/until some other dependency changes (where the dependency must be referenced by `constraint` or `required`, or possibly a form language change affecting `jr:constraintMsg`/`jr:requiredMsg`).

I considered doing something narrower, to subscribe _only_ to the node's `value` state, but I don't think that's really necessary (unless it would add clarity).

- - -

It also occured to me, since I mentioned `constraint` self-references, that we may want a special case for that in the dependency lookup. But it looks like that already exists as of #135.

It's also worth mentioning that there's precedent for self-reference in the spec (and `constraint` is specifically called out in [JavaRosa's DAG explainer](https://github.com/getodk/javarosa/blob/45a73a15bb7901236c73c31a0dc21c945bce2f09/dag.md) as being separate from the main computation DAG).

Reply:
got it.